### PR TITLE
Cray bugfix: TERM missing while reading default target

### PR DESCRIPTION
**Pull request by becker33**
Bug: Spack hangs on some Cray machines

Reason: The `TERM` environment variable is necessary to run `bash -lc "echo $CRAY_CPU_TARGET"`, but we run that command within `env -i`, which wipes the environment.

Fix: Manually forward the `TERM` environment variable to `env -i /bin/bash -lc "echo $CRAY_CPU_TARGET"`

### CODE REVIEW

File: lib/spack/spack/platforms/cray.py [hunk 1]
```diff
@@ -118,7 +118,8 @@ def _default_target_from_env(self):
         # env -i /bin/bash -lc echo $CRAY_CPU_TARGET 2> /dev/null
         if getattr(self, 'default', None) is None:
             env = which('env')
-            output = env("-i", "/bin/bash", "-lc", "echo $CRAY_CPU_TARGET",
+            output = env("-i", "TERM=%s" % os.environ['TERM'],
+                         "/bin/bash", "-lc", "echo $CRAY_CPU_TARGET",
```
Comment (tgamblin) [inline lib/spack/spack/platforms/cray.py]: Can we do this in a slightly less complex way without relying on `env`?

```python
    bash = Executable("/bin/bash")
    bash("-lc", "echo", "$CRAY_CPU_TARGET",
        env={"TERM": os.environ["TERM"]})
```

File: lib/spack/spack/platforms/cray.py [hunk 2]
```diff
@@ -61,7 +61,7 @@ def __init__(self):
                 _target = os.environ.get('SPACK_' + name.upper())
             if _target is None and name == 'back_end':
                 _target = self._default_target_from_env()
-            if _target is not None:
+            if _target:
```
Comment (tgamblin) [inline lib/spack/spack/platforms/cray.py]: What real case is this change handling?

Comment (becker33) [reply lib/spack/spack/platforms/cray.py]: `_default_target_from_env` can return whitespace.

Comment (tgamblin) [reply lib/spack/spack/platforms/cray.py]: I think we should fix that in `_default_target_from_env` instead of here.

Comment (becker33) [reply lib/spack/spack/platforms/cray.py]: I think this is clearer and more pythonic. I went back and forth on this.

`_default_target_from_env` runs a bash command and parses the output. If it gets anything, that's the target, if not, additional logic.

Comment (tgamblin) [reply lib/spack/spack/platforms/cray.py]: I do not think `_default_target_from_env` should return whitespace for failure -- can it really return whitespace *or* `None`?

Comment (becker33) [reply lib/spack/spack/platforms/cray.py]: It can't return `None` currently. It has to return a string.

Comment (becker33) [reply lib/spack/spack/platforms/cray.py]: Reworked so it returns `None`, but I don't think this is an improvement.

**Pull request merged**